File: codes/matlab/ch04/convergence_comparison.m
%% convergence_comparison.m
%
% Compares the convergence of polynomial interpolation on equispaced vs
% Chebyshev nodes for the Runge function.
%
% For smooth functions with singularities at distance d from [-1,1]:
%   - Chebyshev: error ~ O(rho^{-N}) where rho > 1
%   - Equispaced: error can GROW with N
%
% The Runge function has poles at ±0.2i, giving a critical rho ≈ 1.04.
%
% Author: Dr. Denys Dutykh
%         Mathematics Department
%         Khalifa University of Science and Technology
%         Abu Dhabi, UAE
%
% Part of the book "Computational Études: A Spectral Approach"

clear; close all; clc;

%% Configuration
N_FINE = 2000;

% Book color scheme
NAVY = [20, 45, 110] / 255;
SKY = [120, 150, 210] / 255;
CORAL = [231, 76, 60] / 255;
TEAL = [22, 160, 133] / 255;

% Output path
script_dir = fileparts(mfilename('fullpath'));
output_dir = fullfile(script_dir, '..', '..', '..', 'textbook', 'figures', 'ch04', 'matlab');
output_file = fullfile(output_dir, 'convergence_comparison.pdf');

%% The Runge function
runge = @(x) 1 ./ (1 + 25*x.^2);

%% Evaluation grid
x_fine = linspace(-1, 1, N_FINE)';

%% Range of polynomial degrees
N_values = 2:2:50;

errors_equi = zeros(size(N_values));
errors_cheb = zeros(size(N_values));

for i = 1:length(N_values)
    N = N_values(i);

    % Equispaced
    x_equi = linspace(-1, 1, N+1)';
    f_equi = runge(x_equi);
    p_equi = lagrange_interp(x_equi, f_equi, x_fine);
    errors_equi(i) = max(abs(runge(x_fine) - p_equi));

    % Chebyshev
    j = 0:N;
    x_cheb = cos(j * pi / N)';
    f_cheb = runge(x_cheb);
    p_cheb = lagrange_interp(x_cheb, f_cheb, x_fine);
    errors_cheb(i) = max(abs(runge(x_fine) - p_cheb));
end

%% Theoretical convergence rate
% Runge function poles at ±0.2i
rho = abs(0.2i + sqrt(-1.04));

%% Create figure
fig = figure('Units', 'inches', 'Position', [1, 1, 8, 5.5]);
hold on; box on;

% Plot errors
semilogy(N_values, errors_equi, 'o-', 'Color', CORAL, 'LineWidth', 1.5, ...
         'MarkerSize', 5, 'DisplayName', 'Equispaced nodes');
semilogy(N_values, errors_cheb, 's-', 'Color', TEAL, 'LineWidth', 1.5, ...
         'MarkerSize', 5, 'DisplayName', 'Chebyshev nodes');

% Theoretical rate for Chebyshev
idx_fit = N_values >= 20;
C_fit = mean(errors_cheb(idx_fit) .* rho.^N_values(idx_fit));
N_theory = linspace(10, 50, 100);
semilogy(N_theory, C_fit * rho.^(-N_theory), '--', 'Color', TEAL, 'LineWidth', 1.5, ...
         'DisplayName', sprintf('Theory: $O(\\rho^{-N})$, $\\rho \\approx %.2f$', rho));

% Reference line
yline(1, ':', 'Color', [0.5, 0.5, 0.5], 'LineWidth', 0.5, ...
      'DisplayName', 'Error = 1 (useless)');

xlabel('Polynomial degree $N$', 'Interpreter', 'latex');
ylabel('Maximum interpolation error $\|f - p_N\|_\infty$', 'Interpreter', 'latex');
title('Convergence: Equispaced vs Chebyshev Interpolation', 'FontSize', 11);
xlim([0, 52]);
ylim([1e-15, 1e4]);
legend('Location', 'northeast', 'Interpreter', 'latex');

% Annotations
text(30, 1e3, 'Divergence!', 'FontSize', 10, 'Color', CORAL);
text(35, 1e-8, 'Convergence', 'FontSize', 10, 'Color', TEAL);

% Formula box
text(0.05, 0.05, '$f(x) = \frac{1}{1 + 25x^2}$', 'Units', 'normalized', ...
     'Interpreter', 'latex', 'FontSize', 12, ...
     'BackgroundColor', 'white', 'EdgeColor', NAVY, 'LineWidth', 0.5);

set(gca, 'FontSize', 10, 'TickLabelInterpreter', 'latex');
grid on;
set(gca, 'GridAlpha', 0.2);

%% Save figure
if ~exist(output_dir, 'dir')
    mkdir(output_dir);
end

exportgraphics(fig, output_file, 'ContentType', 'vector');
fprintf('Figure saved to: %s\n', output_file);

png_file = strrep(output_file, '.pdf', '.png');
exportgraphics(fig, png_file, 'Resolution', 300);

%% Print summary
fprintf('\nConvergence Comparison (Runge function):\n');
fprintf('%s\n', repmat('-', 1, 55));
fprintf('%4s %15s %15s %12s\n', 'N', 'Equispaced', 'Chebyshev', 'Ratio');
fprintf('%s\n', repmat('-', 1, 55));
for i = 1:5:length(N_values)
    N = N_values(i);
    err_e = errors_equi(i);
    err_c = errors_cheb(i);
    ratio = err_e / err_c;
    fprintf('%4d %15.2e %15.2e %12.1e\n', N, err_e, err_c, ratio);
end
fprintf('%s\n', repmat('-', 1, 55));

fprintf('\nAnalysis:\n');
fprintf('  Runge function poles at z = +/- 0.2i\n');
fprintf('  Chebyshev convergence parameter: rho = %.4f\n', rho);
fprintf('  Expected rate: error ~ rho^{-N} = %.4f^N\n', 1/rho);

% Estimate rho from data
log_errors = log(errors_cheb(10:end));
N_fit = N_values(10:end);
p = polyfit(N_fit, log_errors, 1);
rho_estimated = exp(-p(1));
fprintf('  Estimated from data: rho ~ %.4f\n', rho_estimated);

%% Lagrange interpolation function
function p = lagrange_interp(x_nodes, f_nodes, x_eval)
    n = length(x_nodes);
    p = zeros(size(x_eval));

    for k = 1:n
        L_k = ones(size(x_eval));
        for j = 1:n
            if j ~= k
                L_k = L_k .* (x_eval - x_nodes(j)) / (x_nodes(k) - x_nodes(j));
            end
        end
        p = p + f_nodes(k) * L_k;
    end
end

File: codes/julia/ch04/convergence_comparison.jl
# convergence_comparison.jl
#
# Compares the convergence of polynomial interpolation on equispaced vs
# Chebyshev nodes for the Runge function.
#
# The interpolation error bound is:
#     ||f - p_N||_inf <= (1 + Lambda_N) E_N(f)
#
# where:
#     - Lambda_N is the Lebesgue constant
#     - E_N(f) is the best polynomial approximation error
#
# For the Runge function f(x) = 1/(1 + 25x^2) with poles at +/-0.2i:
#     - Chebyshev: error ~ O(rho^{-N}) where rho ~ 1.04
#     - Equispaced: diverges due to 2^N growth of Lambda_N
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using LaTeXStrings
using Statistics

# -----------------------------------------------------------------------------
# Publication-quality CairoMakie configuration
# -----------------------------------------------------------------------------
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
const N_FINE = 2000

# Book color scheme
const NAVY  = colorant"#142D6E"
const SKY   = colorant"#7896D2"
const CORAL = colorant"#E74C3C"
const TEAL  = colorant"#16A085"

# Output path
const SCRIPT_DIR = @__DIR__
const OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch04", "julia")

# -----------------------------------------------------------------------------
# Functions
# -----------------------------------------------------------------------------
runge(x) = 1.0 / (1.0 + 25.0 * x^2)
equispaced_nodes(N) = collect(range(-1, 1, length=N+1))
chebyshev_nodes(N) = [cos(j * π / N) for j in 0:N]

function lagrange_interpolate(x_nodes, f_nodes, x_eval)
    n = length(x_nodes)
    p_eval = zeros(length(x_eval))

    for k in 1:n
        L_k = ones(length(x_eval))
        for j in 1:n
            if j != k
                L_k .*= (x_eval .- x_nodes[j]) ./ (x_nodes[k] - x_nodes[j])
            end
        end
        p_eval .+= f_nodes[k] .* L_k
    end

    return p_eval
end

function max_interpolation_error(node_func, N; n_eval=N_FINE)
    x_eval = collect(range(-1, 1, length=n_eval))
    x_nodes = node_func(N)
    f_nodes = runge.(x_nodes)
    p_interp = lagrange_interpolate(x_nodes, f_nodes, x_eval)
    return maximum(abs.(runge.(x_eval) .- p_interp))
end

# -----------------------------------------------------------------------------
# Main visualization
# -----------------------------------------------------------------------------
function main()
    # Range of polynomial degrees
    N_values = 2:2:50

    # Compute errors
    errors_equi = [max_interpolation_error(equispaced_nodes, N) for N in N_values]
    errors_cheb = [max_interpolation_error(chebyshev_nodes, N) for N in N_values]

    # Create figure
    fig = Figure(size = (800, 550))
    ax = Axis(fig[1, 1],
              xlabel = L"N" * " (polynomial degree)",
              ylabel = L"\Vert f - p_N\Vert_\infty" * " (max interpolation error)",
              title = "Convergence: Equispaced vs Chebyshev Interpolation",
              limits = ((0, 52), (1e-6, 1e4)),
              yscale = log10)

    # Plot errors
    scatterlines!(ax, collect(N_values), errors_equi, color = CORAL,
                  linewidth = 1.5, markersize = 8, marker = :circle,
                  label = "Equispaced nodes")
    scatterlines!(ax, collect(N_values), errors_cheb, color = TEAL,
                  linewidth = 1.5, markersize = 8, marker = :rect,
                  label = "Chebyshev nodes")

    # Add theoretical convergence rate for Chebyshev
    rho = abs(0.2im + sqrt(complex(-1.04)))
    N_theory = collect(range(10, 50, length=100))

    # Fit the constant
    idx_fit = collect(N_values) .>= 20
    if any(idx_fit)
        C_fit = mean(errors_cheb[idx_fit] .* rho .^ collect(N_values)[idx_fit])
        lines!(ax, N_theory, C_fit .* rho .^ (-N_theory),
               linestyle = :dash, color = TEAL, linewidth = 1.5, alpha = 0.5,
               label = latexstring("Theory: O(\\rho^{-N}), \\rho \\approx $(round(rho, digits=2))"))
    end

    # Reference line
    hlines!(ax, 1, color = colorant"#888888", linewidth = 0.5, linestyle = :dot,
            label = "Error = 1 (useless)")

    # Annotations
    text!(ax, 30.0, 1e3, text = "Divergence!", fontsize = 10, color = CORAL)
    text!(ax, 32.0, 1e-4, text = "Convergence", fontsize = 10, color = TEAL)

    # Formula box
    text!(ax, 0.05, 0.05, text = L"f(x) = \frac{1}{1 + 25x^2}",
          fontsize = 12, space = :relative, align = (:left, :bottom))

    axislegend(ax, position = :rt, framevisible = false, bgcolor = (:white, 0.9))

    # Save figure
    mkpath(OUTPUT_DIR)
    save(joinpath(OUTPUT_DIR, "convergence_comparison.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "convergence_comparison.png"), fig, px_per_unit = 4)

    println("Figure saved to: ", joinpath(OUTPUT_DIR, "convergence_comparison.pdf"))

    # Print summary table
    println("\nConvergence Comparison (Runge function):")
    println("-" ^ 55)
    println("$(rpad("N", 4)) $(lpad("Equispaced", 15)) $(lpad("Chebyshev", 15)) $(lpad("Ratio", 12))")
    println("-" ^ 55)
    for i in 1:5:length(N_values)
        N = N_values[i]
        err_e = errors_equi[i]
        err_c = errors_cheb[i]
        ratio = err_c > 0 ? err_e / err_c : Inf
        println("$(lpad(N, 4)) $(lpad(string(round(err_e, sigdigits=2)), 15)) $(lpad(string(round(err_c, sigdigits=2)), 15)) $(lpad(string(round(ratio, sigdigits=2)), 12))")
    end
    println("-" ^ 55)

    # Analysis
    println("\nAnalysis:")
    println("  Runge function poles at z = +/-0.2i")
    println("  Chebyshev convergence parameter: rho = $(round(rho, digits=4))")
    println("  Expected rate: error ~ rho^{-N} = $(round(1/rho, digits=4))^N")

    # Verify convergence rate
    if length(errors_cheb) > 5
        log_errors = log.(errors_cheb[6:end])
        N_fit = collect(N_values)[6:end]
        # Linear fit in log space
        A = hcat(ones(length(N_fit)), Float64.(N_fit))
        coeffs = A \ log_errors
        rho_estimated = exp(-coeffs[2])
        println("  Estimated from data: rho = $(round(rho_estimated, digits=4))")
    end

    return fig
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
